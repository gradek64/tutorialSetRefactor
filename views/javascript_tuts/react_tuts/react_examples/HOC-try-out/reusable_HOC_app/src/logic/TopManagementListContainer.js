import React from 'react';
import PropTypes from 'prop-types';
import EnhancedList from '../functional/ListComponents/EnhancedList';
import Simplebutton from '../functional/Simplebutton';
import ToogleButton from '../functional/Togglebutton';
import DropDown from '../functional/DropDown';
import SearchBox from '../functional/SearchBox';

/** *********

    This component is like a wrapper it holds all partials componenets as searchBox etc (functionality components)
    the only reason is a class component cause it needs logic for setting filtering, searching etc
    U need to specify handlers here for SearchCompoennt, Sorting, filtering basically anything dynamic user changes to
    the table, If user wasnt interacing with it then it should be functional component not a class.

************ */
const styles = {
  list: {
    listStyleType: 'none',
    backgroundColor: 'white',
    margin: 0,
    padding: 0,
  },
};
const withSimpleState = classes => (EnhancedListComp) => {
  {
    /* the class below is needed for table logic as search, filters etc... */
  }
  class TableStates extends React.Component {
    constructor(props) {
      super(props);
      this.state = {
        sortBy: 'sortASC',
        action: 'sortBy',
        searchValue: '',
        filter: '',
      };
      this.searchText = this.searchText.bind(this);
      this.filterSide = this.filterSide.bind(this);
      this.sortBy = this.sortBy.bind(this);
    }
    searchText(searchValue) {
      this.setState(state => ({
        ...state,
        action: 'search',
        searchValue: searchValue.toLowerCase(),
      }));
    }
    filterSide(filterby) {
      this.setState(state => ({
        ...state,
        filter: filterby,
        action: 'filter',
      }));
    }
    sortBy(sortValue) {
      this.setState(state => ({
        ...state,
        sortBy: sortValue,
        action: 'sortBy',
      }));
    }
    render() {
      return (
        <div>
          <div className="tableHeader">
            <h1>active filter: {this.state.filter === '' ? 'none' : this.state.filter}</h1>
            <h1>active sortBy: {this.state.sortBy}</h1>
            <SearchBox placeholder="filter" eventHandler={this.searchText} />
            <Simplebutton
              eventHandler={this.filterSide}
              value="dark"
              label="filter DARK side"
            />
            <Simplebutton
              eventHandler={this.filterSide}
              value="light"
              label="filter LIGHT side"
            />
            <ToogleButton
              action={this.sortBy}
              label="sort"
              toggleLabels={['ASC', 'DESC']}
              values={['sortASC', 'sortDESC']}
            />
            <div
              style={{
                width: '10%',
                display: 'inline-block',
                backgroundColor: 'white',
              }}
            >
              {/* DropDown is composed with Togglebutton that
                requires eventHandler so U need to specify an empty method */}
              <DropDown label="DropDown">
                <ul style={classes.list}>
                  <li>
                    <Simplebutton
                      eventHandler={this.sortBy}
                      value="sortDESC"
                      label="sortDESC"
                    />
                  </li>
                  <li>
                    {' '}
                    <Simplebutton
                      eventHandler={this.sortBy}
                      value="sortASC"
                      label="sortASC"
                    />
                  </li>
                </ul>
              </DropDown>
            </div>
          </div>
          <div>
            <EnhancedListComp
              {...this.props}
              letterSearch={this.state.searchValue}
              sortBy={this.state.sortBy}
              filter={this.state.filter}
              actionType={this.state.action}
            />
          </div>
        </div>
      );
    }
  }

  TableStates.propTypes = {
    list: PropTypes.instanceOf(Array).isRequired,
  };
  return TableStates;
};

const FilteredListSearch = withSimpleState(styles)(EnhancedList);

export default FilteredListSearch;
