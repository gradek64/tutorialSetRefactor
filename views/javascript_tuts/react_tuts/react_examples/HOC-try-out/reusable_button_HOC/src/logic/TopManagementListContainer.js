import React from "react";
import EnhancedList from "../functional/ListComponents/EnhancedList";
import Simplebutton from "../functional/Simplebutton";
import ToogleButton from "../functional/Togglebutton";
import SearchBox from "../functional/SearchBox";

/** *********

    To make all components reusable again, we need to move the logic back to the [.. withTransformProps] HoC.
    To achieve this, we need to figure out how to use the [...withTransformProps]
    and [...withSimpleState] HoCs with the [...base component] at the same time and
    allow the [...renderDisplayList] to only be responsible for the presentation again.
    We can achieve this behavior using &&&& composition $$$$$.


************ */

/*
  *@this component is responsible for logic only
  *@it has state functionality pass down to Base Component
*/
const withSimpleState = EnhancedListComp => {
  {
    /* the class below is needed for table logic as search, filters etc... */
  }
  return class TableStates extends React.Component {
    constructor(props) {
      super(props);
      this.state = {
        sortBy: "sortASC",
        action: "sortBy",
        searchValue: "",
        filter: ""
      };
      this.searchText = this.searchText.bind(this);
      this.filterSide = this.filterSide.bind(this);
      this.sortBy = this.sortBy.bind(this);
    }
    searchText(searchValue) {
      this.setState(
        state => ({
          ...state,
          action: "search",
          searchValue: searchValue.toLowerCase()
        }),
        () => {
          console.log("this.state", this.state);
        }
      );
    }
    filterSide(filterby) {
      this.setState(state => ({
        filter: filterby,
        ...state,
        action: "filter"
      }));
    }
    sortBy(sortValue) {
      this.setState(state => ({
        sortBy: sortValue,
        ...state,
        action: "sortBy"
      }));
    }
    render() {
      return (
        <React.Fragment>
          <div className="tableHeader">
            <SearchBox placeholder="filter" eventHandler={this.searchText} />
            <Simplebutton
              action={this.filterSide}
              value="dark"
              label="filter DARK side"
            />
            <Simplebutton
              action={this.filterSide}
              value="light"
              label="filter LIGHT side"
            />
            <ToogleButton
              action={this.sortBy}
              label="toggle"
              stateLabels={["in", "out"]}
            />
          </div>
          <div>
            <EnhancedListComp
              {...this.props}
              letterSearch={this.state.searchValue}
              sortBy={this.state.sortBy}
              filter={this.state.filter}
              action={this.state.action}
            />
          </div>
        </React.Fragment>
      );
    }
  };
};

const FilteredListSearch = withSimpleState(EnhancedList);

export default FilteredListSearch;
