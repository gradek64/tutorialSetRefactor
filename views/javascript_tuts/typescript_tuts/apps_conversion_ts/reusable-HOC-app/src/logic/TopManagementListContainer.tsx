import React from 'react';
//import PropTypes from 'prop-types'; no need for those anymore since we check types
import EnhancedList from '../functional/ListComponents/EnhancedList';
import Simplebutton from '../functional/Simplebutton/index';
import ToogleButton from '../functional/Togglebutton/index';
import DropDown from '../functional/DropDown/index';
import SearchBox from '../functional/SearchBox/index';

/** *********

    This component is like a wrapper it holds all partials componenets as searchBox etc (functionality components)
    the only reason is a class component cause it needs logic for setting filtering, searching etc
    U need to specify handlers here for SearchCompoennt, Sorting, filtering basically anything dynamic user changes to
    the table, If user wasnt interacing with it then it should be functional component not a class.

************ */
const styles = {
  list: {
    listStyleType: 'none',
    backgroundColor: 'white',
    margin: 0,
    padding: 0,
  },
};
interface IProps {
  // list: {}[]; //same as 
  list: Array<{name:string,side:string}>; 
}
interface IState {
  sortBy: string;
  action: string;
  searchValue: string;
  filter: string;
}
interface IPropsList extends IProps{
  letterSearch:string;
  sortBy:string;
  filter:string;
  actionType:string;
}

//this is component interface the same as: (prop:IPropsList) => JSX.Element
interface EnhancedListCompInterface {
  (prop:IPropsList): JSX.Element
}

const withSimpleState = (classes:any) => (EnhancedListComp:EnhancedListCompInterface)=> {
  {
    /* the class below is needed for table logic as search, filters etc... */
  }

  // the clock's state has one field: The current time, based upon the
// React class interface



//TableStates class interface
//class itself
interface ITableStates {
  searchText:(searchValue:string)=>void;
  filterSide:(filterby:string)=>void;
  sortBy:(sortValue:string)=>void;
}

  class TableStates extends React.Component<IProps,IState> implements ITableStates {
    /* 
      no need for that anymore:
    
    constructor(props: IProps) {
      super(props);
      this.state = {
        sortBy: 'sortASC',
        action: 'sortBy',
        searchValue: '',
        filter: '',
    } 
      this.searchText = this.searchText.bind(this);
      this.filterSide = this.filterSide.bind(this);
      this.sortBy = this.sortBy.bind(this); 
    */
      greg:string;
      state:Readonly<IState> = {
        sortBy: 'sortASC',
        action: 'sortBy',
        searchValue: '',
        filter: '',
    }
    //searchText = (searchValue:string) => {
    public searchText = (searchValue:string) => {
      this.setState(state => ({
        ...state,
        action: 'search',
        searchValue: searchValue.toLowerCase(),
      }));
    }
    //filterSide = (filterby:string) => {
    public filterSide = (filterby:string) => {
      this.setState(state => ({
        ...state,
        filter: filterby,
        action: 'filter',
      }));
    }
    //sortBy = (sortValue:string) => {
    public sortBy = (sortValue:string) => {
      this.setState(state => ({
        ...state,
        sortBy: sortValue,
        action: 'sortBy',
      }));
    }
    render() {
      return (
        <div>
          <div className="tableHeader">
            <h1>active filter: {this.state.filter === '' ? 'none' : this.state.filter}</h1>
            <h1>active sortBy: {this.state.sortBy}</h1>
            <SearchBox placeholder="filter" eventHandler={this.searchText} />
            <Simplebutton
              eventHandler={this.filterSide}
              value="dark"
              label="filter DARK side"
            />
            <Simplebutton
              eventHandler={this.filterSide}
              value="light"
              label="filter LIGHT side"
            />
            <ToogleButton
              action={this.sortBy}
              label="sort"
              toggleLabels={['ASC', 'DESC']}
              values={['sortASC', 'sortDESC']}
            />
            <div
              style={{
                width: '10%',
                display: 'inline-block',
                backgroundColor: 'white',
              }}
            >
              {/* DropDown is composed with Togglebutton that
                requires eventHandler so U need to specify an empty method */}
              <DropDown label="DropDown">
                <ul style={classes.list}>
                  <li>
                    <Simplebutton
                      eventHandler={this.sortBy}
                      value="sortDESC"
                      label="sortDESC"
                    />
                  </li>
                  <li>
                    {' '}
                    <Simplebutton
                      eventHandler={this.sortBy}
                      value="sortASC"
                      label="sortASC"
                    />
                  </li>
                </ul>
              </DropDown>
            </div>
          </div>
          <div>
            <EnhancedListComp
              {...this.props}
              letterSearch={this.state.searchValue}
              sortBy={this.state.sortBy}
              filter={this.state.filter}
              actionType={this.state.action}
            />
          </div>
        </div>
      );
    }
  }

  /* TableStates.propTypes = {
    list: PropTypes.instanceOf(Array).isRequired,
  }; */
  return TableStates;
};

const FilteredListSearch = withSimpleState(styles)(EnhancedList);

export default FilteredListSearch;
