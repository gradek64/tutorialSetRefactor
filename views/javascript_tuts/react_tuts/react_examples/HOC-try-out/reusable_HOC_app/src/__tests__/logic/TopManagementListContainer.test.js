import React from 'react';
import { shallow } from 'enzyme';
import TopManagementListContainer from '../../logic/TopManagementListContainer';
import { noReactFragementSupport } from '../../test_helpers';

const starWarsChars = [
  { name: 'Luke', side: 'light' },
  { name: 'Darth Vader', side: 'dark' },
];

let wrapper;
beforeAll(() => {
  wrapper = shallow(<TopManagementListContainer list={starWarsChars} />);
});

test('It renders correctly', () => {
  // const wrapperUpdate = noReactFragementSupport(wrapper.debug());
  expect(wrapper).toMatchSnapshot();
});

/*
  *@testing class methods by calling them
  *@directly U dont need to know how and who is calling them in first place
  as buttons, events etc...
*/

test('updates state after calling searchText method', () => {
  // const wrapperUpdate = noReactFragementSupport(wrapper.debug());
  const value = 'V';
  // test classInstance.searchText();
  wrapper.instance().searchText(value);

  const updatedState = wrapper.instance().state;
  expect(updatedState.searchValue).toBe('v');
  expect(updatedState.action).toBe('search');
});

test('updates state after calling sortBy method', () => {
  // test classInstance.sortBy();
  let side = 'ASC';
  wrapper.instance().sortBy(side);

  const updatedState = wrapper.instance().state;
  expect(updatedState.sortBy).toBe(side);
  expect(updatedState.action).toBe('sortBy');

  side = 'DESC';
  wrapper.instance().sortBy(side);

  const updatedStateFilter = wrapper.instance().state;
  expect(updatedStateFilter.sortBy).toBe(side);
  expect(updatedStateFilter.action).toBe('sortBy');
});

test('updates state after calling filterSide method', () => {
  // test classInstance.filerSide();
  let side = 'dark';
  wrapper.instance().filterSide(side);

  const updatedState = wrapper.instance().state;
  expect(updatedState.filter).toBe(side);
  expect(updatedState.action).toBe('filter');

  side = 'light';
  wrapper.instance().filterSide(side);

  const updatedStateFilter = wrapper.instance().state;
  expect(updatedStateFilter.filter).toBe(side);
  expect(updatedStateFilter.action).toBe('filter');
});

/*
  *@U can also test if you render jsx has change after state change
  *@enzyme comes with wrapper.update() and re-renders the jsx again
*/
test('It changes render outcome once filter is changed', () => {
  wrapper.update();
  expect(wrapper.find('h1').at(0).text()).toEqual('active filter: light');
  expect(wrapper.find('h1').at(1).text()).toEqual('active sortBy: DESC');
});

