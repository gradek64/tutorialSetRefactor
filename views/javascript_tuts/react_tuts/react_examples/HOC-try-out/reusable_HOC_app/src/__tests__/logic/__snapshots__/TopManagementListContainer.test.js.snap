// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`It renders correctly 1`] = `
<div>
  <div
    className="tableHeader"
  >
    <h1>
      active filter: 
      none
    </h1>
    <h1>
      active sortBy: 
      sortASC
    </h1>
    <SearchBox
      eventHandler={[Function]}
      placeholder="filter"
    />
    <SimpleButton
      action={[Function]}
      label="filter DARK side"
      value="dark"
    />
    <SimpleButton
      action={[Function]}
      label="filter LIGHT side"
      value="light"
    />
    <ToggleSimpleButton
      action={[Function]}
      label="sort"
      toggleLabels={
        Array [
          "ASC",
          "DESC",
        ]
      }
      values={
        Array [
          "sortASC",
          "sortDESC",
        ]
      }
    />
    <div
      style={
        Object {
          "backgroundColor": "white",
          "display": "inline-block",
          "width": "10%",
        }
      }
    >
      <ToggleSimpleButton
        label="DropDown"
      >
        <ul
          style={
            Object {
              "backgroundColor": "white",
              "listStyleType": "none",
              "margin": 0,
              "padding": 0,
            }
          }
        >
          <li>
            <SimpleButton
              action={[Function]}
              label="sortDESC"
              value="sortDESC"
            />
          </li>
          <li>
             
            <SimpleButton
              action={[Function]}
              label="sortASC"
              value="sortASC"
            />
          </li>
        </ul>
      </ToggleSimpleButton>
    </div>
  </div>
  <div>
    <withTransformPropsComp
      action="sortBy"
      filter=""
      letterSearch=""
      list={
        Array [
          Object {
            "name": "Luke",
            "side": "light",
          },
          Object {
            "name": "Darth Vader",
            "side": "dark",
          },
        ]
      }
      sortBy="sortASC"
    />
  </div>
</div>
`;
